design-documents/distributed-deployment.md

```diff
@@ -0,0 +1,217 @@
+        "type": "project",
+        "require": {
+            "vendor/catalog-extension-admin-ui": "*"
+        },
+        "installed-packages": [
+            "vendor/catalog-extension"
+        ]
+    }
+    ```
+    
+    This would allow to understand which packages that are installed on the instance are part of extension we want to remove and remove these packages from all instances.
+    
+    Pros
+     * The same mechanism can be used for distributed deployment of services
+    
+    Cons
+    * Will require us to publish more projects
+    * Need to add new type of configuration
+    
+    Examples of using command
+    
+    ```
+    instance-manager install --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager uninstall --package="vendor-name/catalog-customization" --instances "./admin-ui ./ui ./cron ./webapi"
+    instance-manager update --instances "./admin-ui ./ui ./cron ./webapi"
+    ```
+    
+    Some of corner cases tool need to support
+    1. When upgrading extension that shares common package with another extension, make sure that version of the common package satisfies metapackage of another package
+    2. When removing package from one of the instances that is part of the extension user want to remove, check that package is not part of any other installed extension
+    3. When updating packages, check that new versions of the packages satisfy version requirements of metapackage
+    4. When updating extension find new versions of extensions that consist from packages that have their dependency/versions met (dependencies on Magento packages). When calculating versions, we probably will have to take into account Magento package versions on other instances
+    
+    #### Tool architecture
+    There are two options for tool implementation
+    
+    1. Tool will be responsible for version calculation and constraint checks.
+    2. To perform operation, tool will be performing this operation on monoliths first and then updating instances using packages on monoliths as a master. Because of the complexity described above this approach might be more reliable.
+
+3. Convention based
+
+    We can have convention similar to this
+    * Modules that end with *admin-ui go to admin instance
+    * Modules that end with *ui and not *admin-ui go to store front instance
+    
+    Cons
+    * Convention can't be reliable option (developers may not follow it, someone may call storefront module login-as-admin)
+    * Will not work for distributed deployment of services
+
+
+4. Introduce new type of configuration, but don't distribute separate projects
+
+    There would be a tool that splits monolith before deployment.
+
+    Pros
+    * Easy upgrade
+    * Easy management of obsolete dependencies
+    
+    Cons
+    * Need to add new type of configuration 
+    * More complicated deployment process
+
+## Open questions
+1. What to do with web setup wizard
```

Comment:
Probably we need to describe a flow how Magento application might be installed with split modules.

Reply:
I mean application for managing modules. I'm not sure we need to support if for distributed deployment.